File: cplusplus/src/autograph.cpp
#include "internal.h"

namespace autograph {

Party create_initiator(const SignFunction sign,
                       const Bytes identity_public_key) {
  return create_party(true, sign, identity_public_key);
}

Party create_initiator(const KeyPair identity_key_pair) {
  return create_initiator(create_sign(identity_key_pair.private_key),
                          identity_key_pair.public_key);
}

Party create_responder(const SignFunction sign,
                       const Bytes identity_public_key) {
  return create_party(false, sign, identity_public_key);
}

Party create_responder(const KeyPair identity_key_pair) {
  return create_responder(create_sign(identity_key_pair.private_key),
                          identity_key_pair.public_key);
}

KeyPair create_key_pair() {
  Bytes private_key(32);
  Bytes public_key(32);
  KeyPair key_pair = {private_key, public_key};
  return key_pair;
}

KeyPairResult generate_ephemeral_key_pair() {
  auto key_pair = create_key_pair();
  bool success = autograph_key_pair_ephemeral(key_pair.private_key.data(),
                                              key_pair.public_key.data()) == 0;
  KeyPairResult result = {success, key_pair};
  return result;
}

KeyPairResult generate_identity_key_pair() {
  auto key_pair = create_key_pair();
  bool success = autograph_key_pair_identity(key_pair.private_key.data(),
                                             key_pair.public_key.data()) == 0;
  KeyPairResult result = {success, key_pair};
  return result;
}

bool init() { return autograph_init() == 0; }

}  // namespace autograph

File: rust/src/autograph.rs
use crate::clib::autograph_init;
use crate::key_exchange::perform_key_exchange;
use crate::key_pair::{generate_ephemeral_key_pair, generate_identity_key_pair};
use crate::safety_number::calculate_safety_number;
use crate::sign::create_sign;
use crate::types::{AutographError, Bytes, KeyExchangeResult, KeyPair, SignFunction};

#[non_exhaustive]
#[derive(Clone, Debug)]
pub struct Autograph;

impl Autograph {
    pub fn new() -> Result<Self, AutographError> {
        if unsafe { autograph_init() } < 0 {
            Err(AutographError::InitializationError)
        } else {
            Ok(Self)
        }
    }

    pub fn calculate_safety_number(&self, a: &Bytes, b: &Bytes) -> Result<Bytes, AutographError> {
        calculate_safety_number(a, b)
    }

    pub fn create_sign<'a>(&'a self, identity_private_key: &'a Bytes) -> SignFunction {
        create_sign(identity_private_key)
    }

    pub fn generate_ephemeral_key_pair(&self) -> Result<KeyPair, AutographError> {
        generate_ephemeral_key_pair()
    }

    pub fn generate_identity_key_pair(&self) -> Result<KeyPair, AutographError> {
        generate_identity_key_pair()
    }

    pub fn perform_key_exchange<'a>(
        &'a self,
        sign: &'a SignFunction,
        our_identity_key: &'a Bytes,
        is_initiator: bool,
        our_ephemeral_key_pair: KeyPair,
        their_identity_key: &'a Bytes,
        their_ephemeral_key: Bytes,
    ) -> KeyExchangeResult {
        perform_key_exchange(
            sign,
            our_identity_key,
            is_initiator,
            our_ephemeral_key_pair,
            their_identity_key,
            their_ephemeral_key,
        )
    }
}
